From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 21pages <sunboeasy@gmail.com>
Date: Tue, 10 Dec 2024 14:12:01 +0800
Subject: [PATCH] videotoolbox changing bitrate

Signed-off-by: 21pages <sunboeasy@gmail.com>
---
 libavcodec/videotoolboxenc.c | 42 +++++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/libavcodec/videotoolboxenc.c b/libavcodec/videotoolboxenc.c
index 0000000000..1111111111 100644
--- a/libavcodec/videotoolboxenc.c
+++ b/libavcodec/videotoolboxenc.c
@@ -289,6 +289,8 @@ typedef struct VTEncContext {
     int power_efficient;
     int max_ref_frames;
     int spatialaq;
+
+    int64_t last_bit_rate;
 } VTEncContext;
 
 static void vtenc_free_buf_node(BufNode *info)
@@ -1193,6 +1195,7 @@ static int vtenc_create_encoder(AVCodecContext   *avctx,
     int64_t      one_second_value = 0;
     void         *nums[2];
 
+    vtctx->last_bit_rate = bit_rate;
     int status = VTCompressionSessionCreate(kCFAllocatorDefault,
                                             avctx->width,
                                             avctx->height,
@@ -2530,6 +2533,45 @@ out:
     return status;
 }
 
+static void update_config(AVCodecContext *avctx)
+{
+    VTEncContext *vtctx = avctx->priv_data;
+
+    if (avctx->codec_id == AV_CODEC_ID_PRORES)
+        return;
+
+    if (avctx->bit_rate == vtctx->last_bit_rate)
+        return;
+
+    av_log(avctx, AV_LOG_INFO, "Setting bit rate to %"PRId64"\n", avctx->bit_rate);
+    vtctx->last_bit_rate = avctx->bit_rate;
+
+    SInt32 bit_rate = avctx->bit_rate;
+    CFNumberRef bit_rate_num = CFNumberCreate(kCFAllocatorDefault,
+                                              kCFNumberSInt32Type,
+                                              &bit_rate);
+    if (!bit_rate_num)
+        return;
+
+    if (vtctx->constant_bit_rate) {
+        int status = VTSessionSetProperty(vtctx->session,
+                                          compat_keys.kVTCompressionPropertyKey_ConstantBitRate,
+                                          bit_rate_num);
+        if (status == kVTPropertyNotSupportedErr) {
+            av_log(avctx, AV_LOG_ERROR, "-constant_bit_rate true is not supported by the encoder.\n");
+        }
+    } else {
+        int status = VTSessionSetProperty(vtctx->session,
+                                          kVTCompressionPropertyKey_AverageBitRate,
+                                          bit_rate_num);
+        if (status) {
+            av_log(avctx, AV_LOG_ERROR, "Error: cannot set average bit rate: %d\n", status);
+        }
+    }
+
+    CFRelease(bit_rate_num);
+}
+
 static av_cold int vtenc_frame(
     AVCodecContext *avctx,
     AVPacket       *pkt,
@@ -2542,6 +2584,7 @@ static av_cold int vtenc_frame(
     CMSampleBufferRef buf = NULL;
     ExtraSEI sei = {0};
 
+    update_config(avctx);
     if (frame) {
         status = vtenc_send_frame(avctx, vtctx, frame);
 
-- 
2.43.0
